File: Solutions/1658. Minimum Operations to Reduce X to Zero/1658. Minimum Operations to Reduce X to Zero.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int x) {
        int target=-x;
        for (auto& num:nums){
            target+=num;
        }
        if(target==0){
            return nums.size();
        }
        unordered_map<int,int> m;
        m[0]=-1;
        int sum=0;
        int res=INT_MIN;

        for(int i=0;i<nums.size();++i){
            sum+=nums[i];
            if(m.count(sum-target)) {
                res=max(res,i-m[sum-target]);
            }
            m[sum]=i;
        }
        return res==INT_MIN?-1:nums.size()-res;
    }
};

File: Solutions/1658. Minimum Operations to Reduce X to Zero/1658. Minimum Operations to Reduce X to Zero.go
func minOperations(nums []int, x int) int {
    target:=-x;
    for _,num:=range nums{
        target+=num
    }
    if target==0{
        return len(nums)
    }
    m:=make(map[int]int)
    m[0]=-1;
    sum,res:=0,math.MinInt

    for i:=0;i<len(nums);i+=1{
        sum+=nums[i];
        if _,ok:=m[sum-target];ok{
            res=max(res,i-m[sum-target]);
        }
        m[sum]=i;
    }
    if res==math.MinInt{
        return -1
    }else{
        return len(nums)-res
    }
}

func max(i,j int) int{
    if i>j{
        return i
    }
    return j
}